ibtk/src/lagrangian/FEDataManager.cpp

```diff
@@ -1144,6 +1146,122 @@ FEDataManager::prolongData(const int f_data_idx,
     return;
 } // prolongData
 
+/**
+ * The three functions below allow compile-time selection of the number of
+ * basis functions and number of variables. Fixing the loop bounds via
+ * templates and switch statements makes most IBFE codes a few percent faster.
+ *
+ * This function (the last of the three) takes the number of variables and
+ * number of basis functions as template argument which allows the compiler to
+ * unroll the two inner loops when appropriate. Setting either value to -1
+ * results in run-time selection of loop bounds (which is considerably
+ * slower).
+ */
+template <int n_vars, int n_basis>
+void
+integrate_cell_rhs_n_vars_n_basis(const unsigned int qp_offset,
+                                  const std::vector<std::vector<double> >& phi_F,
```

Comment:
If we want to go even further down the rabbit hole when we rewrite FEDataInterpolation it would be great to replace `std::vector<std::vector<double>>` with a contiguous 2D array where the basis functions are indexed fastest (since we always index by quadrature points slowest).

Reply:
Yes indeed...

Reply:
Also, let's not call this a `cell`. Maybe `elem`? I would like to try to reserve `cell` for the background grid cells.